examples/benchmark.py

```diff
@@ -31,183 +124,469 @@ def downloadAmberSuite():
     steps = 20
     while True:
-        time = timeIntegration(context, steps, initialSteps)
-        if time >= 0.5*options.seconds:
+        elapsed_time = timeIntegration(context, steps, initialSteps)
+        if elapsed_time >= 0.5*options.seconds:
             break
-        if time < 0.5:
-            steps = int(steps*1.0/time) # Integrate enough steps to get a reasonable estimate for how many we'll need.
+        if elapsed_time < 0.5:
+            steps = int(steps*1.0/elapsed_time) # Integrate enough steps to get a reasonable estimate for how many we'll need.
+        else:
+            steps = int(steps*options.seconds/elapsed_time)
+    test_result['steps'] = steps
+    test_result['elapsed_time'] = elapsed_time
+    time_per_step = elapsed_time * unit.seconds / steps 
+    ns_per_day = (integ.getStepSize() / time_per_step) / (unit.nanoseconds/unit.day)
+    test_result['ns_per_day'] = ns_per_day
+
+    # Serialize XML files for Folding@home benchmark if requested
+    if options.serialize:
+        wu_duration = 5*unit.minutes
+        ncheckpoints_per_wu = 2
+        nsteps_per_wu = int(wu_duration / time_per_step)
+
+        coredata = dict()
+        coredata['checkpointFreq'] = int(nsteps_per_wu / ncheckpoints_per_wu)
+        coredata['numSteps'] = ncheckpoints_per_wu * coredata['checkpointFreq']
+        coredata['xtcFreq'] = coredata['numSteps']
+        coredata['precision'] = options.precision
+        coredata['xtcAtoms'] = 'solute'
+        coredata['disableCheckpointStateTests'] = 1 # disable checkpoint state tests
+        coredata['DisablePmeStream'] = 0 # don't disable separate PME streams
+        coredata['forceTolerance'] = 200000 # for some reason, we have to make this large (TODO: Investigate why)
+        coredata['energyTolerance'] = 200000 # for some reason, we have to make this large (TODO: Investigate why)
+        
+        forces = { force.getName() : force for force in system.getForces() }
+        NONBONDED_METHODS = { 0 : 'NoCutoff', 1 : 'CutoffNonPeriodic', 2 : 'CutoffPeriodic', 3 : 'Ewald', 4 : 'PME', 5 : 'LJPME' }
+        if 'NonbondedForce' in forces:
+            nonbonded_method = NONBONDED_METHODS[forces['NonbondedForce'].getNonbondedMethod()]
         else:
-            steps = int(steps*options.seconds/time)
-    print('Integrated %d steps in %g seconds' % (steps, time))
-    print('%g ns/day' % (dt*steps*86400/time).value_in_unit(unit.nanoseconds))
+            nonbonded_method = 'unknown'
+
+        metadata = {
+            'name' : testName,
+            'description' : f'OpenMM Benchmark Suite : {testName}',
+            'precision' : options.precision,
+            'num_atoms' : context.getSystem().getNumParticles(),
+            'nonbonded_method' : f'{nonbonded_method}',
+            'timestep' : integ.getStepSize() / unit.picoseconds,
+            'integrator' : f'{integ.__class__.__name__}',
+            }
```

Comment:
Does all of this really belong here?  Serializing the system and integrator is fine, but this is quite a lot of F@H specific code being put into OpenMM.

Reply:
We're aiming to benchmark each release of OpenMM on all available GPUs, which requires serializing out all the benchmarks to a new Folding@home project each release. This is the easiest way to do it without maintaining separate benchmarking scripts, which runs the risk of having to keep both in perfect sync to be a valid comparison.

This is just an example script anyway, so there's no harm in keeping it here if it serves an active purpose each release and within the development cycles.